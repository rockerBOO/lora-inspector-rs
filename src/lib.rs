use std::fmt::Debug;
// use std::alloc::Global;
use std::string::String;

// use candle_core::safetensors;
// use std::error::Error;
// use std::marker::Send;
// use std::marker::Sync;

mod file;
mod metadata;
mod network;
mod norms;
mod weight;
mod worker;

pub type Result<T> = std::result::Result<T, InspectorError>;

#[derive(Debug)]
pub enum InspectorError {
    Candle(candle_core::Error),
    SafeTensor(safetensors::SafeTensorError),
    Load(String),
    Msg(String),
}

impl InspectorError {
    fn candle(err: candle_core::Error) -> InspectorError {
        InspectorError::Candle(err)
    }

    fn safetensor(err: safetensors::SafeTensorError) -> InspectorError {
        InspectorError::SafeTensor(err)
    }
}

impl From<candle_core::Error> for InspectorError {
    fn from(err: candle_core::Error) -> InspectorError {
        InspectorError::candle(err)
    }
}

impl From<safetensors::SafeTensorError> for InspectorError {
    fn from(err: safetensors::SafeTensorError) -> InspectorError {
        InspectorError::safetensor(err)
    }
}

pub fn get_base_name(name: &str) -> String {
    name.split('.')
        .filter(|part| !matches!(*part, "weight" | "lora_up" | "lora_down" | "alpha"))
        .fold(String::new(), |acc, v| {
            if acc.is_empty() {
                v.to_owned()
            } else {
                format!("{acc}.{v}")
            }
        })
}

#[cfg(test)]
mod tests {
    use crate::get_base_name;

    #[test]
    fn get_base_name_test() {
        let base_name = get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.lora_up.weight");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");

        let base_name =
            get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.lora_down.weight");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");

        let base_name = get_base_name("lora_unet_up_blocks_1_attentions_1_proj_out.alpha");
        assert_eq!(base_name, "lora_unet_up_blocks_1_attentions_1_proj_out");
    }
}
