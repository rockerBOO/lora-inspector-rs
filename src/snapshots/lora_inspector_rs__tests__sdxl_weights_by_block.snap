---
source: src/lib.rs
expression: dbg!(result)
---
{
  "text_encoder": {
    "layer_11": 2.236038566105992,
    "layer_03": 2.559074085055627,
    "layer_09": 2.605124137412777,
    "layer_15": 2.9749570401214847,
    "layer_18": 2.9165044534280433,
    "layer_06": 2.516811396222816,
    "layer_25": 2.7996566374047074,
    "layer_31": 3.8637997801833954,
    "layer_30": 2.8715523784561747,
    "layer_26": 2.90682780986022,
    "layer_02": 2.585883574878324,
    "layer_17": 2.856573554443203,
    "layer_20": 2.7940594038519677,
    "layer_04": 2.4886652477788647,
    "layer_19": 2.9181846132605895,
    "layer_01": 2.4362615302873563,
    "layer_21": 2.8049301858088875,
    "layer_28": 2.906921097006807,
    "layer_27": 2.6981724264439197,
    "layer_23": 2.861240965268156,
    "layer_24": 2.827270217244079,
    "layer_10": 2.582895048458408,
    "layer_08": 2.5526433805076016,
    "layer_29": 2.798727564344977,
    "layer_05": 2.4642735594170753,
    "layer_00": 2.3986237865212283,
    "layer_16": 2.9225455991238536,
    "layer_13": 2.864615594965109,
    "layer_07": 2.5771845546871783,
    "layer_14": 2.82534215634467,
    "layer_12": 2.9098976914799457,
    "layer_22": 2.9790958653316753
  },
  "unet": {
    "input_00": 3.93684786919843,
    "middle_01": 3.710052128450166,
    "middle_00": 4.49129544650758,
    "middle_02": 4.209200318584085,
    "middle_03": 7.846870343021595,
    "output_03": 5.532903021129883,
    "output_02": 4.571775593484964,
    "output_00": 4.158135129624979,
    "input_02": 3.78792256247538,
    "input_03": 4.395878374398332,
    "input_01": 3.2558379804519544,
    "output_01": 3.7485393705952395
  }
}
